src/core/finalize.ts

```diff
@@ -43,163 +47,280 @@ export function processResult(result: any, scope: ImmerScope) {
+		if (!state || !isSameScope(state, rootScope)) {
 			return
 		}
-		if (
-			parentState &&
-			parentState.base_ &&
-			parentState.base_[prop] === childValue &&
-			childIsFrozen
-		) {
-			// Object is unchanged from base - no need to process further
-			return
+
+		// Handle potential set value finalization first
+		fixPotentialSetContents(state)
+
+		const finalizedValue = getFinalValue(state)
+
+		// Update all locations in the parent that referenced this draft
+		updateDraftInParent(parent, state.draft_ ?? state, finalizedValue, key)
+
+		generatePatchesAndFinalize(state, patches, inversePatches)
+	})
+}
+
+function generatePatchesAndFinalize(
+	state: ImmerState,
+	patches?: Patch[],
+	inversePatches?: Patch[]
+) {
+	const shouldFinalize =
+		state.modified_ &&
+		!state.finalized_ &&
+		(state.type_ === ArchType.Set ||
+			(state.assigned_ && state.assigned_.size > 0))
+
+	if (shouldFinalize) {
+		if (patches && inversePatches) {
+			const patchPlugin = getPlugin("Patches")
+			const basePath = patchPlugin.getPath(state)
+
+			if (basePath) {
+				patchPlugin.generatePatches_(state, basePath, patches, inversePatches)
+			}
+		}
+
+		markStateFinalized(state)
+	}
+}
+
+export function handleCrossReference(
+	target: ImmerState,
+	key: string | number | symbol,
+	value: any
+) {
+	// Check if value is a draft from this scope
+	if (isDraft(value)) {
+		const state: ImmerState = value[DRAFT_STATE]
+		if (isSameScope(state, target.scope_)) {
+			// Register callback to update this location when the draft finalizes
+
+			state.callbacks_.push(function crossReferenceCleanup() {
+				// Update the target location with finalized value
+				prepareCopy(target)
+
+				const finalizedValue = getFinalValue(state)
+
+				updateDraftInParent(target, value, finalizedValue, key)
+			})
 		}
-		finalize(rootScope, childValue)
-		// Immer deep freezes plain objects, so if there is no parent state, we freeze as well
-		// Per #590, we never freeze symbolic properties. Just to make sure don't accidentally interfere
-		// with other frameworks.
-		if (
-			(!parentState || !parentState.scope_.parent_) &&
-			typeof prop !== "symbol" &&
-			(isMap(targetObject)
-				? targetObject.has(prop)
-				: Object.prototype.propertyIsEnumerable.call(targetObject, prop))
+	} else if (isDraftable(value)) {
+		// Handle non-draft objects that might contain drafts
+		target.callbacks_.push(function nestedDraftCleanup() {
+			const targetCopy = latest(target)
+
+			if (get(targetCopy, key, target.type_) === value) {
+				// Process the value to replace any nested drafts
+				finalizeAssigned(target, key, target.scope_)
+			}
+		})
+	}
+}
+
+export function finalizeAssigned(
+	state: ImmerState,
+	key: PropertyKey,
+	rootScope: ImmerScope
+) {
+	const wasAssigned =
+		(state as Exclude<ImmerState, SetState>).assigned_!.get(key) ?? false
+
+	if (rootScope.drafts_.length > 1 && wasAssigned === true && state.copy_) {
```

Comment:
NIT: wasAssigned looks like the most expensive condition, so maybe inline and put the copy_ check upfront?